src/base/RandomUuid.cc

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 1996-2022 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+
+#include "base/IoManip.h"
+#include "base/RandomUuid.h"
+#include "base/TextException.h"
+#include "defines.h"
+
+#include <iostream>
+#include <random>
+
+static_assert(sizeof(RandomUuid) == 128/8, "RandomUuid has RFC 4122-prescribed 128-bit size");
+
+RandomUuid::RandomUuid()
+{
+    // Generate random bits for populating our UUID.
+    // STL implementation bugs notwithstanding (e.g., MinGW bug #338), this is
+    // our best change of getting a non-deterministic seed value for the r.n.g.
+    static std::random_device dev; // unknown a priori size (sizeof(int)) values
+    static std::mt19937_64 rng(dev()); // known 64-bit size values
+    const auto rnd1 = rng();
+    const auto rnd2 = rng();
+
+    // bullet 3 of RFC 4122 Section 4.4 algorithm but setting _all_ bits (KISS)
```

Comment:
@yadij, suggesting a specific comment text that you want to see would be very helpful here! Please keep in mind that the PR code implements the UUID format prescribed by the RFC, including the total number of random bits (122 IIRC). In other words, there is no "loss" or "gain" of collision chances here compared to what the RFC design implies!

Any general-purpose RNG should continue to supply 128 bits of (total) randomness here for as long as we memcpy() those random bits like we do now. The existing static assertions will check that assumption (to the extent they can).

If we incorrectly promise (somewhere) that these UUIDs have 128 bits of randomness, then _that_ text should be adjusted to remove that false promise.

The RFC algorithm suggests setting the 6 "fixed" bits first and then setting the other 122 bits to random values. Instead, we set all 128 bits first and then overwrite the 6 "fixed" ones just because it is a lot simpler than setting the right 122 random bits (while preserving the already set 6). Those 6 "fixed" bits are not located at the beginning or end of the UUID memory... The existing comments already document the change of RFC steps order and the reason for that change.

To avoid misunderstanding, I do not see a need to document more details here, so I cannot suggest specific wording to address this change request, but I am _not_ against adding a (correct) comment detailing something that others think should be detailed, of course.

Reply:
The algorithm specified by the RFC guarantees uniqueness up to the limit of whatever the RNG chosen.

The way you have simplified has side effects. Specifically the discarding of randomness bits in fixed positions reduces that randomness. Front or back or elsewhere is irrelevant.

```suggestion
    // bullet 3 of RFC 4122 Section 4.4 algorithm but setting _all_ bits (KISS).
    // This reduces the Mersenne-Twister 2^19937 cycle to an equivalent of 2^19002
    // (i.e. values before collisions are expected) which is still sufficient for this use-case.
```

Reply:
> The algorithm specified by the RFC guarantees uniqueness up to the limit of whatever the RNG chosen.

I am not familiar with the term "RNG limit", but if you mean that the probability of v4 UUID collisions depends on the quality of the chosen RNG, then I naturally agree.
 
> The way you have simplified has side effects. 

There are no negative side effects here that I know about as far as general "randomness" of produced UUID values is concerned. AFAICT, the effects you are thinking about are not actually relevant to produced UUID values, as detailed below.

> Specifically the discarding of randomness bits in fixed positions reduces that randomness.

Fortunately, "that randomness" is irrelevant here; only the randomness that actually goes into the UUID values is relevant. For example, imagine that we generate a perfectly random number one million bits in size and then take just the 122 bits (of those one million bits) that UUIDs actually need/use. Are we significantly "reducing that [one million bit] randomness" by only looking at 122 bits out of one million? For some definition of "randomness", yes, we are! Is that in any way relevant to the quality of the produced UUID? Not at all! The quality of the produced UUID would be the same whether we start with one million random bits, one trillion random bits, or 128 random bits. What matters is that we end up using exactly 122 of those bits.

Discarding bits at fixed positions does not reduce the randomness of the remaining bits. Given a good random number generator every fixed bit has the same "randomness" -- `rnd & 1` has the same "randomness" as `rnd & 16` bit has the same "randomness" as `rnd & 256` (where `&` is the bitwise "and" operator).

Defects in bad random number generators may prevent folks from using their generated value bits freely, but std::mt19937_64 randomness quality is good enough to avoid those concerns, especially in the context of the PR implementation where 122 out of 128 bits remain.

> Front or back or elsewhere is irrelevant.

True. I mentioned front/back/elsewhere to explain why the PR uses this particular implementation/simplification. It has no effect on the randomness of the result (i.e. on the UUID uniqueness, cycle, period, etc.).

> `This reduces the Mersenne-Twister 2^19937 cycle to an equivalent of 2^19002`

My math skills are insufficient to quickly come up with correct numbers, but numbers 2<sup>19937</sup> and 2<sup>19002</sup> do not define the period of random values in UUIDs produced by the PR code, neither before the 6-bit truncation, nor after that truncation. Perhaps more importantly, the comment incorrectly implies that we are artificially reducing the randomness of produced UUIDs. We do not.

Thus, I am against adding the proposed comment.

> `2^19002 (i.e. values before collisions are expected)`

Collisions are expected _before_ the period ends -- despite the word "Unique" in the name, random (i.e. v4) UUIDs are not guaranteed to be unique. [Wikipedia](https://en.wikipedia.org/wiki/Universally_unique_identifier#Collisions) estimates v4 UUID collision probability within 103 trillion v4 UUIDs as "one in a billion" (but that estimate probably assumes perfect randomness, so the real probability will be higher).

When the random number generator period actually ends, collision probability becomes 100% because the random number generator starts producing exactly the same numbers. However, in our context, that will never happen because a Squid process will never run for such a long time.

Reply:
>  will reduce the randomness 

I agree with Alex that setting 6 bits with predefined values cannot reduce radnomness of other 122 bits. Basically (with the given seed), Mersenne Twister will keep producing random bits (i.e. values with the same randomness) until reaching its huge 2<sup>1937</sup> period. We can [select](https://stackoverflow.com/questions/471157/is-a-subset-of-a-random-sequence-also-random) any subset of these bits, and it should also contain a random sequence. In this respect I do not see a connection of taking away 6 bits (in the produced result) and the supposed reducing of the period to 2^19002.

Reply:
If you want me to teach you probability theory of randomness vs distribution bias I can try, but that will add delay to this PR.

All I am just asking is to add the comment to signal that the effect of ignoring these bits has been considered and determined adequate for this use-case. Any changes to the RNG in future will need to re-do that consideration.

Reply:
> If you want me to teach you probability theory of randomness vs distribution bias I can try

FWIW, I do not want you to try that. To make better progress, I suggest assuming that others know (or can discover) what you know as far as probability theory is concerned and use that knowledge to prove your statements and advance your argument. Using standard terminology is always a good idea. Stating key assumptions may be helpful as well.

AFAICT, there is a disconnect between what you know and the PR code, but I do not know where that misunderstanding is rooted yet, so I can only check specific statements for correctness (which I did).

> All I am just asking is to add the comment to signal that the effect of ignoring these bits has been considered and determined adequate for this use-case. Any changes to the RNG in future will need to re-do that consideration.

I am still not against adding a (valid) comment here. The specific comment you suggested is invalid -- it misleads and lies about what is actually going on in this code (as I have tried to detail/explain [earlier](https://github.com/squid-cache/squid/pull/1015#discussion_r843417426)).

Would the following (unformatted) comment address your concern?

```C++
// We generated 128 (equally) random bits but are going to use only 122 of them
// because 6 bits are occupied by the required UUID version and variant info
// that is written during algorithm bullets one and two (implemented below).
```

Reply:
Very well. What the RFC requires is that a UUID is made up of 6 specific-value bits, and 122 bits "set randomly".

What you have implemented is different - a tempered semisimple subset/mapping of the Mersenne-Twister distribution.

The example we give students here is that of a dice roll:

 . Roll a standard 6-sided dice and you can expect a value from {1,2,3,4,5,6}.

 . Roll a 6-sided dice whose values have a binary mask applied (eg bit #2 always is 0). Its **values** now come from the cycle {1, 0, 1, 4, 5, 4} with a set of {0,1,4,5}.

 . Roll a standard 4-sided dice with sides labelled from that set of {0,1,4,5}.

When rolled the **sequence** is still random for all three, but the **value** is "more random" for the standard dice. The probability of any **value** from both the standard 4-side and 6-side is ```P(N)=1/N```. There is a predictable bias to the values when the masked dice is rolled: {1,4} are twice as likely to occur than {0, 5}.
```

What you have implemented in this PR masks 6-bits off the values pulled from the M-T distribution sequence, with a **cycle** of 128-bits. Like the dice example, this erases randomness out of the M-T distribution sequence. A measurable ```(2^((19937/128)*6)``` **values** have just doubled (or more) their probability of occuring. Instead of a ```q=2^19937``` cycle of randomness we get a tempered semisimple sequence with ``u=19937, r=128, m=6 ;; q= 2^(u-(u/r)*m) = 2^19002``` quality randomness.

What the RFC describes (a non-masking approach) is equivalent to the 4-sided dice. A complicated implementation taking 122-bits from the mt19937_64 sequence and leaving/preserving the unused 6-bits for the next UUID to be generated preserved the **quality** of randomness for **values** extracted. Even though the set of those values is necessarily fewer due to 128-bit vs 122-bit size difference.


Again, all I am asking is a comment to highlight that we are aware of this effect and it have determined that the M-T 19937 64-bit distribution/sequence has sufficient **quality** randomness to meet the needs of UUID.

Reply:
TLDR:

> Again, all I am asking is a comment to highlight that we are aware of this masking side-effect and it have determined that the M-T 19937 64-bit distribution/sequence has sufficient **quality** randomness to meet the needs of UUID even when it is used.

The only thing standing in the way of adding a comment is finding a correct description of what you call a "masking side-effect". The comment version you proposed earlier had factual errors. You have not commented on my earlier comment version, but I can propose another version as a reaction to your latest comments discussed below, in hope to make progress (let's call this comment v3 for ease of reference):

```C++
// The code assumes that each of the 128 random bits is independent and
// uniformly distributed, allowing us to use any 122 of them (the remaining 6
// bits are occupied by the required UUID version and variant info that is
// written during algorithm bullets one and two that are implemented below). No
// real r.n.g. is perfect, but we assume that the quality of std::mt19937_64 is
// high enough to make any imperfections irrelevant to this specific code.
```

Would adding the above comment address your concern?

If you want to dive into specific low-level problems with the side-effect analysis, please see below, but I certainly do not insist on continuing that argument if we can agree on a comment that does not depend on that argument outcome.

----


> What the RFC requires is that a UUID is made up of 6 specific-value bits, and 122 bits "set randomly".

Agreed.

> What you have implemented is different

That is a claim you have to prove. So far, you have not done so IMO.


> The example we give students here is that of a dice roll:
> 
> . Roll a standard 6-sided dice and you can expect a value from {1,2,3,4,5,6}.

As you probably know, the example does not apply to the PR code: PR code uses a set of independent uniformly distributed random bits (let's call that "equally random"). The bits in your example are not equally random. The PR code can pick and choose bits at will. The 6-sided dice example cannot.

It is easy to make the example more applicable by using an 8-sided dice with values {0,1,2,3,4,5,6,7}. In that example, picking any subset of bits will preserve the overall properties of the generated random numbers.
 
> . Roll a 6-sided dice whose values have a binary mask applied (eg bit #2 always is 0). Its **values** now come from the cycle {1, 0, 1, 4, 5, 4} with a set of {0,1,4,5}.

BTW, there is no "cycle" when rolling a standard dice (with or without bit manipulations) -- a dice is not a pseudo-random number generator that cycles through a fixed set of values. I am not trying to nitpick here; I mention this problem only because incorrect statements about cycles disqualified the first comment version you have proposed. Please be extra careful if you insist on documenting any specific cycles in the official comment; most of the cycle-related numbers I have seen so far were problematic...
 
> What you have implemented in this PR masks 6-bits off the values pulled from the M-T distribution sequence,

Agreed.

> with a **cycle** of 128-bits.

I am not sure what you mean by a "**cycle** of 128 bits". FWIW, std::mt19937_64, the random number generator used by the PR code has a period close to 2<sup>19937</sup>. The length of that period is not determined by the number of bits in generated integers. The std::mt19937 generator has approximately the same period despite producing 32-bit instead of 64-bit integers!

I believe that the UUID value cycle produced by the PR code is close to 2<sup>19937-1</sup> (i.e. the r.n.g. period divided by 2 because we are using two generated integers for every one generated UUID). If you prefer, we can explicitly document that belief instead of leaving the current "the resulting period is long enough" assumption implicit.

> Like the dice example, this erases randomness out of the M-T distribution sequence. 

"Erasing randomness" is not a standard term, but the PR code does not change the randomness of the random bits it uses. The code assumes that the original 128 bits are uniformly distributed and independent -- each bit distribution has the same properties as another. Thus, the PR code can pick and choose the bits it needs; the used 122 bits remain uniformly distributed and independent. The same cannot be said about the bits in the (inapplicable) 6-sided dice example, of course.

> A measurable `(2^((19937/128)*6)` **values** have just doubled (or more) their probability of occuring. 

This statement is incorrect or misleading. We should only be looking at the used 122 bits or values. Only their properties matter. Among those bits, _nothing_ have changed. They are exactly the same bits that came from the r.n.g. Those 122 bits do not "know" that there were 6 other bits somewhere. All 2<sup>128-6</sup> UUID values produced by this code are equally probable (within the code assumptions regarding the r.n.g. quality, of course).

> Instead of a `q=2^19937` cycle of randomness we get a tempered semisimple sequence with ``u=19937, r=128, m=6 ;; q= 2^(u-(u/r)*m) = 2^19002``` quality randomness.

If you decide to continue this discussion, please provide a reference describing the concept you called a "tempered semisimple sequence" and defining u, r, n, and q. I believe the assertion that 122 bits have a different cycle than 128 bits is incorrect, but I cannot pinpoint the failure without knowing how you got from the 2<sup>19937</sup> period to the 2<sup>19002</sup> period (if that is what you meant).

The period here is determined by the std::mt19937_64 generator. The used 122 bits have pretty much the same period as the original 128 bits. Within our assumptions, even if we were to take a _single_ bit (instead of 122) out of 128, that single bit period will remain the same as the period of the original 128 bits! (Thank you to @eduard-bagdasaryan for pointing that out.)

The preservation of the period can be proven: For simplicity, let's assume that we are dealing with std::mt19937_64 that generates high quality[^1] 64-bit integers on the interval [0, 2<sup>64</sup>), and has a well-known period P. The period of a single bit[^2][^3] is R (`R <= P`). Now we generate and remember P 64-bit integers (i.e. run one whole std::mt19937_64 period). Look at the values of the first bit of those P integers. The sequence of those values will begin to repeat after the first R integers (because the bit period is R). Same for the second bit and so on: `bit k of integer[i] == bit k of integer[i+R]` for any bit position `k` and any integer generation step `i`. Thus, all integers will repeat after R: `integer[i] == integer[i+R]` for any integer generation step `i`. That is the definition of the integers generation period! Thus, `R == P` -- the bit period is P.

[^1]: "High quality" here means that generated integers are uniformly distributed over the entire interval -- no significant hot spots, gaps, and similar first-order anomalies (at least). Needless to say, the actual generator we are dealing with is not perfect in all aspects -- it does have defects -- but we assume that, in the specific context of a RandomUuid() constructor, those imperfections are small enough to preserve the general validity of math models that ignore those defects. Without those (reasonable) assumptions, we cannot make any general statements about the code using the generator.

[^2]: I am omitting reasoning why R must exist and cannot exceed P because I consider those assertions easy to prove, but I can prove them if needed.

[^3]: The period R is the same for all 64 bits because we are dealing with a high quality[^1] generator; significantly different periods would create significant hot spots and other anomalies that high quality generators should not have (just like in the {1,2,3,4,5,6} dice counter-example). Another possible way to prove same-R-for-all-bits assertion is that even a single bit period that is not R would prevent the generator from restarting the whole sequence (essentially return to the initial state) when generating integer number P if that special bit position has not finished its period yet (unless P is divisible by the bit period). Mersenne Twister period is a Mersenne prime number so, P cannot be divisible by the bit period that is not P itself. Perhaps this reasoning can be even morphed into a shorter/simpler main proof!

 
> What the RFC describes (a non-masking approach) is equivalent to the 4-sided dice. 

The "non-masking approach" part of the above statement is misleading. The RFC does not dictate how those random bits are obtained. The RFC does not prohibit any masking as long as that masking does not spoil the randomness of the used bits.

> A complicated implementation taking 122-bits from the mt19937_64 sequence and leaving/preserving the unused 6-bits for the next UUID to be generated preserved the **quality** of randomness for **values** extracted. 

I agree that one could preserve 6, 12, 18, etc. unused bits and use them on the next iteration, but that complication feels irrelevant: The fact that other (more complex) correct implementations exist does not mean that the PR implementation is incorrect. You have not demonstrated the the existing implementation decreases "quality of randomness" (I do not think you have even defined what "quality of randomness" is).


> Even though the set of those values is necessarily fewer due to 128-bit vs 122-bit size difference.

Yes, any high quality implementation will result in the same set of possible UUID values because that set is determined by UUID format, not the underlying generator details (as long as the generator is a high quality one, of course).

Reply:
> TLDR:
> 
> > Again, all I am asking is a comment to highlight that we are aware of this masking side-effect and it have determined that the M-T 19937 64-bit distribution/sequence has sufficient **quality** randomness to meet the needs of UUID even when it is used.
> 
> The only thing standing in the way of adding a comment is finding a correct description of what you call a "masking side-effect".

The mathematical term is "temper". But that is unlikely to mean much to devs. Whereas "mask" (code technical term) and "side-effect" (English and code technical term) are pretty clear and common programming terminology. I'm having a hard time understanding why you are confused by them. Is it that I am using them to describe some advanced maths calculation?


> The comment version you proposed earlier had factual errors.

That allegation has not been supported by any of your responses to it so far - just a misunderstanding on your part of one of the terms I used.

As requested I have supplied the formula and terminology to lookup references with. The proofs themselves are highly complex math going on about equivalencies of mappings across geometry spaces - way beyond me repeating here.

> You have not commented on my earlier comment version,

Ah, okay. My response is that it in no way gives any hints to future dev that there may be something worth investigating if/when they change the mt19937 engine or its seeding.


> but I can propose another version as a reaction to your latest comments discussed below, in hope to make progress (let's call this comment v3 for ease of reference):
> 
> ```c++
> // The code assumes that each of the 128 random bits is independent and
> // uniformly distributed, allowing us to use any 122 of them (the remaining 6
> // bits are occupied by the required UUID version and variant info that is
> // written during algorithm bullets one and two that are implemented below). No
> // real r.n.g. is perfect, but we assume that the quality of std::mt19937_64 is
> // high enough to make any imperfections irrelevant to this specific code.
> ```
> 
> Would adding the above comment address your concern?
> 

This subset reads to me like what I am trying to have mentioned:
 ```c++
 // No real r.n.g. is perfect, but we assume that the quality of std::mt19937_64 is
 // high enough to make any imperfections irrelevant to this specific code.
 ```

As a team "we" can do more than "assume" as well, even if "you" are assuming. The math I ran shows the **quality** stays very large.


> If you want to dive into specific low-level problems with the side-effect analysis, please see below, but I certainly do not insist on continuing that argument if we can agree on a comment that does not depend on that argument outcome.
> 

If the effect was significant I would have made this request a blocker and done so. But it is (relatively) small, so no. Just a "simple" (sigh) highlight comment and move on (or so I thought).


> > What the RFC requires is that a UUID is made up of 6 specific-value bits, and 122 bits "set randomly".
> 
> Agreed.
> 
> > What you have implemented is different
> 
> That is a claim you have to prove. So far, you have not done so IMO.
> 

My claim is demonstrated by the implemented code. The 122-bits are set via masking a random number. This is a non-random algorithm applying a "temper" to the RNG.

This is a side-track no relevant to the issue at hand though, which is the quality of randomness in those 122-bits used.


> > The example we give students here is that of a dice roll:
> > . Roll a standard 6-sided dice and you can expect a value from {1,2,3,4,5,6}.
> 
> As you probably know, the example does not apply to the PR code: PR code uses a set of independent uniformly distributed random bits (let's call that "equally random"). The bits in your example are not equally random. The PR code can pick and choose bits at will. The 6-sided dice example cannot.

The dice example picks bit #2. You can pick any other bit(s) you like to mask away from the dice. All possible non-0 bit mask reduces the randomness of the output below the quality of the generator (dice).
Unfortunately this trivial example is great for demo but terrible for showing the 2^N calculations at work. I just have to trust the experts myself on that connection.

> 
> It is easy to make the example more applicable by using an 8-sided dice with values {0,1,2,3,4,5,6,7}. In that example, picking any subset of bits will preserve the overall properties of the generated random numbers.
> 
> > . Roll a 6-sided dice whose values have a binary mask applied (eg bit #2 always is 0). Its **values** now come from the cycle {1, 0, 1, 4, 5, 4} with a set of {0,1,4,5}.
> 
> BTW, there is no "cycle" when rolling a standard dice (with or without bit manipulations) -- a dice is not a pseudo-random number generator that cycles through a fixed set of values.

This maybe a good example of the terms difference. Dice roll has a cycle of 2log(6) bits because that is the binary length of the distinct number of "sides" you can roll to. Dice have no defined period that I am aware of - maybe infinite as a truly random source. All that is certain is that it is at least 2log(6).

> I am not trying to nitpick here; I mention this problem only because incorrect statements about cycles disqualified the first comment version you have proposed. Please be extra careful if you insist on documenting any specific cycles in the official comment; most of the cycle-related numbers I have seen so far were problematic...
> 

It means the number of repetitions of some pattern occurs. There are two types of cycle that could be referred to in this discussion. One is the repeating nature of the M-T sequence - however, that is carefully being called "period" [except in my original suggestion text]. The other is the cycle of 6-per-128 bits masked away by this PR implementation - I have tried to be consistent only calling that "cycle" since we started having to talk about its specifics.

Cycle can retain the same *length* as the period (AFAICT that is what your math proves).  But the values retrievable have less randomness (i.e. bias added) due to the masking operation.

In this PR case the mask erases 6-bit of every 128-bit extracted. Those bits are gone forever.

Take it to the extremes:

A) mask away 0 bits and you will have the full 2^19937 random bits available.

B) mask away 127 bits and you will have only ```2^(19937-(19937/128)*6) ~= 2^155``` random bits across the entire period.

C) mask away 128 bits and you will have a constant 0.

So we get my original suggestion - despite using an RNG which promises 2^19937 random bits, this PR mask operation throws away 2^935 of them. Leaving *at most* 2^19002 random bits un-masked. aka:
```
This reduces the Mersenne-Twister 2^19937 ... to an equivalent of 2^19002
```


> > What you have implemented in this PR masks 6-bits off the values pulled from the M-T distribution sequence,
> 
> Agreed.
> 
> > with a **cycle** of 128-bits.
> 
> I am not sure what you mean by a "**cycle** of 128 bits".

That appears to be a major part of your misinterpreting my math. See above, hopefully that clears it up.


> I believe that the UUID value cycle produced by the PR code is close to 219937-1 (i.e. the r.n.g. period divided by 2 because we are using two generated integers for every one generated UUID). If you prefer, we can explicitly document that belief instead of leaving the current "the resulting period is long enough" assumption implicit.
> 

That is another "temper" algorithm being applied to this M-T sequence/period. However, its effect is to halve the available uniqueness randomness (extracting two values), then double it (appending bit-wise from the same sequence). So overall no change to the quality from this step AFAIK.


> > Like the dice example, this erases randomness out of the M-T distribution sequence.
> 
> "Erasing randomness" is not a standard term, but the PR code does not change the randomness of the random bits it uses.

FYI; I **bold** highlighted the terms in my previous post. That was a descriptive phrasing.

You seem so close to understanding there. Almost. Consider that this PR "uses" 128-bits from the M-T bit sequence (aka period). For 6-bits that "use" is to mask them away / erase.


> The code assumes that the original 128 bits are uniformly distributed and independent -- each bit distribution has the same properties as another.

Yes.

> Thus, the PR code can pick and choose the bits it needs

No. The very act of "picking" which random bits are to be used ... changes the distribution. The dice example shows that clearly.


> > A measurable `(2^((19937/128)*6)` **values** have just doubled (or more) their probability of occuring.
> 
> This statement is incorrect or misleading.

> We should only be looking at the used 122 bits or values. 

That is a common misconception. Pertaining to the mt19937 period there are 128 "used" bits in this algorithm.

How those 128 random bits are being mapped onto 122-bit output values is the issue.

> 
> > Instead of a `q=2^19937` cycle of randomness we get a tempered semisimple sequence with ``u=19937, r=128, m=6 ;; q= 2^(u-(u/r)*m) = 2^19002``` quality randomness.
> 
> If you decide to continue this discussion, please provide a reference describing the concept you called a "tempered semisimple sequence" 

[https://en.wikipedia.org/wiki/Tempered_representation](https://en.wikipedia.org/wiki/Tempered_representation)

The M-T algorithm treats its 2^19937 bits as a infinitely repeating sequence - making it confirm to the requirements of a  [https://en.wikipedia.org/wiki/Locally_compact_abelian_group] which can be 


>and defining u, r, n, and q.

They are algebraic symbols defined in the text you quoted.


> I believe the assertion that 122 bits have a different cycle than 128 bits is incorrect, but I cannot pinpoint the failure without knowing how you got from the 219937 period to the 219002 period (if that is what you meant).

It is not.

2^19002 is the maximum number of random bits you can use from mt19937 RNG engine if you throw away 6 of every 128 fetched for each UUID. After that it will (supposed to anyway) give you the same bits as your very first fetch.

2^19937 is the maximum number of bits you can use from mt19937 RNG engine if you extract only 122 bits for each UUID.



> > What the RFC describes (a non-masking approach) is equivalent to the 4-sided dice.
> 
> The "non-masking approach" part of the above statement is misleading. The RFC does not dictate how those random bits are obtained. The RFC does not prohibit any masking as long as that masking does not spoil the randomness of the used bits.

The 6-bit mask applied in this PR breaks exactly that "as long as" condition. Bias the distribution towards producing a certain set of UUID.

> 
> > Even though the set of those values is necessarily fewer due to 128-bit vs 122-bit size difference.
> 
> Yes, any high quality implementation 

Er, "quality" in this topic under discussion is a measurable property of a random number distribution. It has a value that can be calculated with in mathematical formulas. "high quality implementation" refers to code quality yes?

Reply:
> this will reduce the randomness

I tried to compare our implementation with the official RFC4122 sample [implementation](https://datatracker.ietf.org/doc/html/rfc4122#appendix-A), but unfortunately, it does not include our version 4. However, we have Boost UUID library, covering all versions including [V4](https://www.boost.org/doc/libs/1_73_0/boost/uuid/random_generator.hpp). @yadij , please have a look at their basic_random_generator::operator(), which 1) it sets random bits  2) overwrites version bits (our PR does the same). They would have warned about any problems with this approach, including 'randomness reducing', I guess.

Reply:
TLDR: I have committed the part of the comment v4 that Amos seemed to like, with a trivial modification to keep the formatted comment on two 80-character lines (v5 -- commit 2317407):

```c++
// No real r.n.g. is perfect, but we assume that std::mt19937_64 quality is
// high enough to make any imperfections irrelevant to this specific code.
```

The text below is the continuation of the discussion of what is actually going on in the code. During this iteration, we discover a yet-to-be-properly-defined term that may be key to resolving some of the remaining misunderstandings and correct some math mistakes. I certainly do _not_ insist on continuing that part of the discussion.

----

> The mathematical term is "temper". But that is unlikely to mean much to devs. Whereas "mask" (code technical term) and "side-effect" (English and code technical term) are pretty clear and common programming terminology. I'm having a hard time understanding why you are confused by them. 

I am not confused by the terms "temper", "mask", and "side-effect".


> As requested I have supplied the formula and terminology to lookup references with.

You did not supply enough: A lot of sequences can be called "tempered semisimple sequence", a lot of things match a `q=2^(u-(u/r)*m)` formula, and there are many ways to measure "randomness quality". The information you have provided so far is just not enough to understand what you actually mean. I am pretty sure you are not using the right math for what is actually going on in the PR code, but I cannot pinpoint the exact problem because you do not show your steps/proofs and do not provide enough information to properly identify the alleged result of those steps.


> The proofs themselves are highly complex math going on about equivalencies of mappings across geometry spaces - way beyond me repeating here.

It is your call whether to repeat something here, but even a field expert would need a lot more to go on to understand the assertion that we are dealing with a "tempered semisimple sequence with `u=19937, r=128, m=6 ;; q= 2^(u-(u/r)*m) = 2^19002` quality randomness", especially when lots of things can be called "tempered" and "semisimple" -- those terms describe an infinite number of algorithms/mappings and their applicability to any particular formula (or the PR code!) has not been established yet. It is not even clear what you mean by "quality randomness" in this context! If you are doing original research, it is on you to show it. If you are referring to facts established by others, it is on you to provide appropriate references to those facts.


> This subset reads to me like what I am trying to have mentioned:
> 
> ```c++
> // No real r.n.g. is perfect, but we assume that the quality of std::mt19937_64 is
> // high enough to make any imperfections irrelevant to this specific code.
> ```

Great, I have committed the above version (with a trivial modification). Glad I was able to figure out what you were trying to mention! I remain open to adding more information to that comment.


> As a team "we" can do more than "assume" as well, even if "you" are assuming. The math I ran shows the **quality** stays very large.

As a team, we can indeed do more if we cooperate. So far, your conclusions raise red flags, and you have not shown your math in a way that the rest of the team can validate. That alone does not mean you are right or wrong -- the team work is not about individual correctness but about the ability of the team members to improve (and rely on) each other results, moving the team forward. I cannot improve (or rely on) the conclusions you have presented so far, but I am trying hard to get them to a state where that becomes possible.


> > > Amos: What the RFC requires is that a UUID is made up of 6 specific-value bits, and 122 bits "set randomly".

> > Alex: Agreed.

> > > What you have implemented is different

> > That is a claim you have to prove. So far, you have not done so IMO.
 
> Amos: My claim is demonstrated by the implemented code. The 122-bits are set via masking a random number. This is a non-random algorithm applying a "temper" to the RNG.

Tempering can increase, decrease, or leave unchanged various randomness qualities. From pure algebra point of view, "tempering" is just an application of a function or operation. In the context of random sequences, the term is often used to describe functions that _improve_ some quality of the random sequence (as in "tempered glass is better than ordinary glass"). It is on you to show that the PR code decreases some UUID uniqueness quality if you want to make that claim a part of the comment. You cannot just claim that tempering implies a decrease of quality. That implication is false. For example, the Mersenne Twister algorithm itself uses several forms of tempering to _improve_ certain aspects of the resulting sequence!


> > BTW, there is no "cycle" when rolling a standard dice -- a dice is not a pseudo-random number generator that cycles through a fixed set of values.

> Dice roll has a cycle of 2log(6) bits because that is the binary length of the distinct number of "sides" you can roll to. Dice have no defined period that I am aware of - maybe infinite as a truly random source. All that is certain is that it is at least 2log(6).

> It means the number of repetitions of some pattern occurs.

The number of values that repeat at least once across the period of the random number generator? Or the number of those value repetitions? How can either be applicable to a standard dice -- a standard dice does not have a period at all?? Please define that term precisely. I do not think "cycle" is a well-known term in this subject area.

Your `2log(6)` formula does not let me reconstruct the actual "cycle" definition. It looks suspiciously close to a formula for the entropy of a two-dice roll, but we were not (and should not be) discussing entropy so that is probably just a coincidence.


> There are two types of cycle that could be referred to in this discussion. One is the repeating nature of the M-T sequence - however, that is carefully being called "period" [except in my original suggestion text]. The other is the cycle of 6-per-128 bits masked away by this PR implementation - I have tried to be consistent only calling that "cycle" since we started having to talk about its specifics.

Without a definition of a "cycle", I still do not know what you mean. Some bits are removed from _every_ value generated by our M-T generator. There is no "cycle" there in any common sense of that word. I am happy to use your terminology, but you need to define it to make such reuse (and basic understanding!) possible.

We can come back to your other cycle-dependent statements armed with that definition. They are not usable without it.

 
> In this PR case the mask erases 6-bit of every 128-bit extracted. Those bits are gone forever.

Agreed.

> Take it to the extremes:

> A) mask away 0 bits and you will have the full 2^19937 random bits available.

The total number of random bits is not 2<sup>19937</sup>. Each generated number within a generation period has 128 random bits. Since we are generating two random numbers per UUID, there will be P/2 generations. Thus, the total is `128*P/2` or, approximately, 2<sup>19937+6</sup> random bits. 

... where P is std::mt19937_64 period -- approximately 2<sup>19937</sup>.


> B) mask away 127 bits and you will have only `2^(19937-(19937/128)*6) ~= 2^155` random bits across the entire period.

The total number of random bits in this case is P/2 -- one bit per generation -- approximately 2<sup>19937-1</sup>.

I do not know why you are dividing P by 2<sup>((19937/128)*6)</sup> in your formula, but the result of that expression is 2<sup>19007</sup> and not 2<sup>155</sup>. Something went wrong here, multiple times. Please explain your formulas and be more careful to save time on these very expensive iterations!


> C) mask away 128 bits and you will have a constant 0.

To be more precise, there will be no random bits at all for our use in this case; no UUIDs can be generated.


> So we get my original suggestion - despite using an RNG which promises 2^19937 random bits, this PR mask operation throws away 2^935 of them. Leaving _at most_ 2^19002 random bits un-masked. 

Correcting for math mistakes similar to those in items A and B above, the PR code actually uses `(128-6)*P/2` bits, which is more than `32*P` (which is approximately 2<sup>19937+5</sup>). FWIW, I do not know where `2^935` came from -- that number does not appear anywhere else in your comment AFAICT.


> aka: `This reduces the Mersenne-Twister 2^19937 ... to an equivalent of 2^19002`

It does not (as corrected above), but at least I know what you were trying to compute now. It would have saved us a lot of time if you showed your work from the very beginning!

Unfortunately, replacing the above numbers with the correct ones is not going to fully fix that comment. The problem here is that we are not using a 128-bit Mersenne-Twister (while ignoring 6 bits). We are using one 64-bit Mersenne-Twister twice. That _halves_ P as far as UUID generation is concerned and requires a different comment if we are to document that kind of thing.


> > I believe that the UUID value cycle produced by the PR code is close to 219937-1 (i.e. the r.n.g. period divided by 2 because we are using two generated integers for every one generated UUID). If you prefer, we can explicitly document that belief instead of leaving the current "the resulting period is long enough" assumption implicit.
 
> That is another "temper" algorithm being applied to this M-T sequence/period. However, its effect is to halve the available unique randomness (extracting two values), then double it (appending bit-wise from the same sequence). So overall no change to the quality from this step AFAIK.

Whether there is a positive, negative, or zero effect depends on the definition of "quality" (and its object). The generation period is clearly halved -- one can generate 50% fewer UUIDs -- but whether that affects "quality" or not depends on those definitions. Any future comment would need to be careful to avoid or define those words.

 

> > The code assumes that the original 128 bits are uniformly distributed and independent -- each bit distribution has the same properties as another.
> 
> Yes.
> 
> > Thus, the PR code can pick and choose the bits it needs
> 
> No. The very act of "picking" which random bits are to be used ... changes the distribution. The dice example shows that clearly.

"Changing the distribution" does not mean "reduce randomness", for many definitions of "distribution" and "randomness". And the dice example is clearly inapplicable. It only illustrates danger in general.

In general, there are certainly cases where masking bits affects the random qualities we should care about -- one would be right to be very careful/suspicious when doing such things! There are also cases where there is no known negative effect, and even cases where there is a known positive effect. I claim that the PR code is a case where there is no such negative effect (for the reasons I have already documented).

 
> > > Amos: Instead of a `q=2^19937` cycle of randomness we get a tempered semisimple sequence with ``u=19937, r=128, m=6 ;; q= 2^(u-(u/r)*m) = 2^19002``` quality randomness.

> Amos: https://en.wikipedia.org/wiki/Tempered_representation
> 
> The M-T algorithm treats its 2^19937 bits as a infinitely repeating sequence - making it confirm to the requirements of a [https://en.wikipedia.org/wiki/Locally_compact_abelian_group] which can be processed as [https://en.wikipedia.org/wiki/Semisimple_Lie_algebra#Semisimple_and_reductive_groups]

AFAICT, locally compact abelian group is such a vast concept that includes, for example, any finite group of integers. I do not see the relevance of that concept to our specific discussion. Similar things can be said about semisimple and reductive groups -- these are very broad concepts applicable to lots of different algorithms/mappings. I do not know what you mean by "[abelian group] can be processed as semisimple and reductive groups" in this context, but let's assume that it can be. It is still not clear how you get from that to q=2<sup>(u-(u/r)*m)</sup>, and what that `q` and its parameters are. Those critical steps remain a mystery.

> > Alex: and defining u, r, n, and q.

> Amos: They are algebraic symbols defined in the text you quoted.

Their _meaning_ is undefined in that text. To meaningfully apply any formula to a discussion, one needs to know what objects that formula manipulates and computes. These are the very basics of any math-based communication, and they are currently missing.

 
> 2^19002 is the maximum number of random bits you can use from mt19937 RNG engine if you throw away 6 of every 128 fetched for each UUID.

What makes you think that? AFAICT, the total number of random bits in this case is `128-6` multiplied by the number of possible generations `P/2` (i.e. the number of UUID generations before std::mt19937, used twice per UUID generation, reaches the end of its period P) or `(128-6)*P/2` -- a number exceeding 2<sup>19937+5</sup> -- not 2<sup>19002</sup>. 

FWIW, the total number of possible UUID generations is P/2, approximately 2<sup>9968</sup>.

> 2^19937 is the maximum number of bits you can use from mt19937 RNG engine if you extract only 122 bits for each UUID.

That number is approximately `128*P/2` or `2^19937+6`. See corrections for `A)` above.

If we use all available random bits (by preserving instead of discarding some across generations), then the total number of UUID generations will grow from `P/2` to `P/2 + 6*(P/2)/128` or an increase of about 5%. Not worth the trouble when P/2 (i.e. 2<sup>9968</sup>) is already far more than any Squid process can generate -- if my math is right, it would take 2<sup>9913</sup> _years_ to generate `P/2` UUIDs while generating one UUID per nanosecond!


> > Alex: The RFC does not prohibit any masking as long as that masking does not spoil the randomness of the used bits.
 
> Amos: The 6-bit mask applied in this PR breaks exactly that "as long as" condition. Biasing the distribution towards producing a certain set of UUID.

What makes you think that? Not all bit masks introduce bias to the remaining/used bits.

Reply:
> > > Alex: The RFC does not prohibit any masking as long as that masking does not spoil the randomness of the used bits.
> 
> > Amos: The 6-bit mask applied in this PR breaks exactly that "as long as" condition. Biasing the distribution towards producing a certain set of UUID.
> 
> What makes you think that? Not all bit masks introduce bias to the remaining/used bits.

Yes they do, it is the nature of masking vs other operations.

Specific to this PR, the M-T algorithm provides the property of its output values being *evenly distributed* across the value space (1 to 2^64-1). Applying a mask operation breaks that property, causing a (small) bias towards certain UUID being generated consistently more often than others.

Reply:
> Alex: The RFC does not prohibit any masking as long as that masking does not spoil the randomness of the used bits.

> Amos: The 6-bit mask applied in this PR breaks exactly that "as long as" condition. Biasing the distribution towards producing a certain set of UUID.

> Alex: What makes you think that? Not all bit masks introduce bias to the remaining/used bits.

> Amos: Yes they do, it is the nature of masking vs other operations.

I can prove that your assertion is false with a simple example: Consider a random generator library that exports a random number generator called rngAB. Internally, the library takes two independent random number generators, each producing perfect random 1-bit numbers: rngA and rngB. The library code concatenates their output, resulting in generation of two-bit random numbers. The library exports the result as a two-bit random number generator called rngAB. Now, I will use rngAB from that library to produce random numbers, but I will only use the first bit of every number generated by rngAB. The used bit will obviously be perfectly random -- it will be the bit produced by rngA. Thus, masking one bit off did not "introduce bias to the remaining/used bits" in this specific example.

> Specific to this PR, the M-T algorithm provides the property of its output values being _evenly distributed_ across the value space (1 to 2^64-1).

Yes, plus zero: [0, 2<sup>64</sup>-1).

> Applying a mask operation breaks that property, causing a (small) bias towards certain UUID being generated consistently more often than others.

Just to avoid misunderstanding, my statement was not about the original-size values carrying always-zero masked-off bits. My statement was about the _used_ random bits that _remain_ after masking -- the only thing that matters as far as generated UUIDs are concerned. In other words, we must look at the _used_ 122 bits, not the masked 128-bit value (with 6 bits always zero due to our masking).

What makes you think that the random component of UUIDs will not be evenly distributed across its value space (i.e. across [0, 2<sup>122</sup>-1)? If it makes things easier, you can try to prove that the first 60[^1] random bits of generated UUIDs will not be evenly distributed across their value space (i.e. across [0, 2<sup>60</sup>-1) -- those 60 bits come from one MT-generated 64-bit number.

[^1]:  "60 bits" is just an arbitrary number that is small enough to contain sequential random bits from one number produced by the MT generator. You can use 30 bits or even 1 bit if it helps. All those bits should have same first-level randomness properties as far as this discussion is concerned.

Reply:
> > Alex: The RFC does not prohibit any masking as long as that masking does not spoil the randomness of the used bits.
> 
> > Amos: The 6-bit mask applied in this PR breaks exactly that "as long as" condition. Biasing the distribution towards producing a certain set of UUID.
> 
> > Alex: What makes you think that? Not all bit masks introduce bias to the remaining/used bits.
> 
> > Amos: Yes they do, it is the nature of masking vs other operations.
> 
> I can prove that your assertion is false with a simple example:

No. Your example is consistent with my statements, math, and examples given.

> Consider a random generator library that exports a random number generator called rngAB. Internally, the library takes two independent random number generators, each producing perfect random 1-bit numbers: 

By assuming "perfect" random generator your logic requires infinite period for both RNG. Which the dice example already showed the empirical evidence of bias when masking an RNG with infinite period.

By assuming perfection you gave rngA a quality of P=infinityA,  and rngB the matching P=infinityB.

The append operation gave the hypothetical library output an *improved* quality of P= append( infinityA, infinityB). 


> The used bit will obviously be perfectly random -- it will be the bit produced by rngA. 

No. The mask selects only the bits from rngA. Obviously reducing the quality of **library** down to the maximum quality of rngA alone.   P= mask( f( infinityA, infinityB)) = infinityA.


N.B. The fact that we (humanity) cannot determine the exact values of infinityA and infinityB is why the 2^(X...) formulas are not supposed to be applied when "perfect" randomness is present. The proofs show that they still work, but the numbers are meaningless abstractions of infinity. The answers are only of actual use when pseudo-random generators with known/measurable properties are considered.

Reply:
I am resolving this conversation for purposes of the review - my request has been met already. @rousskov if you want to continue the sidetrack discussion of formulas we can continue elsewhere.

Reply:
> By assuming "perfect" random generator your logic requires infinite period for both RNG.

As far as my proof/example are concerned, the RNG period is unused and irrelevant. One can replace "perfect" with any RNG or pseudo-RNG characteristic, and the proof/example will still invalidate the overreaching assertion about masking always introducing bias to used bits. I used "perfect" in hope to make the example more convincing (perfection is often the hardest characteristic to preserve), but it is not material to the example and the related discussion.

> I am resolving this conversation for purposes of the review - my request has been met already.

Thank you.

> @rousskov if you want to continue the sidetrack discussion of formulas we can continue elsewhere.

I regret that I failed to address your underlying concerns (for the lack of a better word) and am ready to continue discussing interpretations of this code effects in whatever format you prefer. I agree that this PR is not the best place for those discussions.